src/rgw/rgw_data_sync.cc

```diff
@@ -5410,148 +5344,396 @@ int RGWSyncBucketCR::operate(const DoutPrefixProvider *dpp)
+		       << ret << dendl;
+    return tl::unexpected(ret);
+  }
+  RGWCoroutinesStack *stack = new RGWCoroutinesStack(store->ctx(), cr_mgr.get());
+  std::vector<rgw_bucket_sync_pair_info> pairs(num_shards);
+  for (auto shard = 0u; shard < num_shards; ++shard) {
+    auto& pair = pairs[shard];
+    pair.source_bs.bucket = sz->info.bucket;
+    pair.dest_bucket = sz->dest;
+    pair.source_bs.shard_id = shard;
+    stack->call(new RGWReadBucketPipeSyncStatusCoroutine(
+		  &sz->sc, pair, &sync_status[shard],
+		  nullptr, latest_gen));
+  }
+
+  stacks.push_back(stack);
+
+  ret = cr_mgr->run(dpp, stacks);
+  if (ret < 0) {
+    ldpp_dout(this, 0) << "ERROR: failed to read sync status for "
+		       << bucket_str{dest_bucket} << dendl;
+    return tl::unexpected(ret);
+  }
 
-    stacks.push_back(stack);
+  return sync_status;
+}
+
+namespace rgw::bucket_sync_run {
+// Retry-loop over calls to sync_bucket_shard_cr
+class ShardCR : public RGWCoroutine {
+  static constexpr auto allowed_retries = 10u;
+
+  RGWDataSyncCtx& sc;
+  const rgw_bucket_sync_pair_info& pair;
+  const uint64_t gen;
+  unsigned retries = 0;
+
+  ceph::real_time prev_progress;
+  ceph::real_time progress;
+
+public:
+
+  ShardCR(RGWDataSyncCtx& sc, const rgw_bucket_sync_pair_info& pair,
+	  const uint64_t gen)
+    : RGWCoroutine(sc.cct), sc(sc), pair(pair), gen(gen) {}
+
+  int operate(const DoutPrefixProvider *dpp) override {
+    reenter(this) {
+      // Since all errors (except ECANCELED) are considered retriable,
+      // retry other errors so long as we're making progress.
+      for (retries = 0u, retcode = -EDOM;
+	   (retries < allowed_retries) && (retcode != 0);
+	   ++retries) {
+	yield call(sync_bucket_shard_cr(&sc, nullptr, pair, gen,
+					sc.env->sync_tracer->root_node,
+					&progress));
+
+	if (retcode == -ECANCELED) {
+	  ldpp_dout(dpp, 0) << "ERROR: Got -ECANCELED for "
+			    << pair.source_bs << dendl;
+	  return set_cr_error(retcode);
+	} else if (retcode < 0) {
+	  ldpp_dout(dpp, 0) << "WARNING: Got error, retcode=" << retcode << " for "
+			    << pair.source_bs << "on retry "
+			    << retries + 1 << " of " << allowed_retries
+			    << " allowed" << dendl;
+	  // Reset the retry counter if we made any progress
+	  if (progress != prev_progress) {
+	    retries = 0;
+	  }
+	  prev_progress = progress;
+	}
+      }
+      if (retcode < 0) {
+	ldpp_dout(dpp, -1) << "ERROR: Exhausted retries for "
+			   << pair.source_bs << " retcode="
+			   << retcode << dendl;
+	return set_cr_error(retcode);
+      }
+
+      return set_cr_done();
+    }
+    return 0;
   }
+};
 
-  int ret = cr_mgr.run(dpp, stacks);
-  if (ret < 0) {
-    ldpp_dout(this, 0) << "ERROR: failed to read sync status for "
-        << bucket_str{dest_bucket} << dendl;
-    return ret;
+// Loop over calls to ShardCR with limited concurrency
+class GenCR : public RGWShardCollectCR {
+  static constexpr auto MAX_CONCURRENT_SHARDS = 16;
```

Comment:
probably worth raising this quite a bit, maybe 64? for comparison, syncing radosgws will spawn up to 20 bucket-shard syncs per datalog shard (128)

eventually we might use rgw_data_sync_spawn_window from https://github.com/ceph/ceph/pull/45784 here, just so it's configurable on the command line